ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,205 @@
+   *                                for every segment. Could be {@code null} if interested only in total valid data size
+   * @return a {@link Pair} of reference time and total valid data size
+   * @throws IOException
+   * @throws StoreException
+   */
+  private Pair<Long, Long> getValidDataSize(SortedMap<String, Long> validDataSizeBySegments)
+      throws IOException, StoreException {
+    long validDataSize = 0;
+    long totalValidDataSize = 0;
+    long referenceTime = time.milliseconds();
+    Iterator<IndexSegment> indexSegmentIterator = index.indexes.values().iterator();
+    IndexSegment indexSegment = null;
+    if (indexSegmentIterator.hasNext()) {
+      indexSegment = indexSegmentIterator.next();
+      validDataSize += getValidDataSizePerIndexSegment(indexSegment, referenceTime);
+    }
+    while (indexSegmentIterator.hasNext()) {
+      IndexSegment nextIndexSegment = indexSegmentIterator.next();
+      if (nextIndexSegment.getLogSegmentName().equals(indexSegment.getLogSegmentName())) {
+        // index segment refers to same log segment as previous index segment
+        validDataSize += getValidDataSizePerIndexSegment(nextIndexSegment, referenceTime);
+      } else {
+        // index segment refers to a new log segment compared to previous index segment
+        String logSegmentName = indexSegment.getLogSegmentName();
+        totalValidDataSize += validDataSize;
+        if (validDataSizeBySegments != null) {
+          validDataSizeBySegments.put(logSegmentName, validDataSize);
+        }
+        indexSegment = nextIndexSegment;
+        validDataSize = getValidDataSizePerIndexSegment(nextIndexSegment, referenceTime);
+      }
+    }
+    if (index.indexes.size() > 1) {
+      String logSegmentName = indexSegment.getLogSegmentName();
+      totalValidDataSize += validDataSize;
+      if (validDataSizeBySegments != null) {
+        validDataSizeBySegments.put(logSegmentName, validDataSize);
+      }
+    }
+    return new Pair<>(referenceTime, totalValidDataSize);
+  }
+
+  /**
+   * Get valid data size of an index segment
+   * @param indexSegment the {@link IndexSegment} for which valid data size has to be determined
+   * @param referenceTimeInMs time in ms used as reference to check for expiration
+   * @return the valid data size of the given index segment
+   * @throws IOException
+   * @throws StoreException
+   */
+  private long getValidDataSizePerIndexSegment(IndexSegment indexSegment, long referenceTimeInMs)
+      throws IOException, StoreException {
+    List<MessageInfo> messageInfos = new ArrayList<>();
+    indexSegment.getEntriesSince(null, new FindEntriesCondition(Integer.MAX_VALUE), messageInfos, new AtomicLong(0));
+    long validSize = 0;
+    for (MessageInfo messageInfo : messageInfos) {
+      IndexValue value = index.findKey(messageInfo.getStoreKey());
+      if (!PersistentIndex.isDeleted(value) && !isExpired(messageInfo.getExpirationTimeInMs(), referenceTimeInMs)) {
```

Comment:
What this method does is
Fetch all entries for the given index segment. 
   For entry check with entire index to understand if the entry is deleted or not
   If not, add it to valid data size

So, we have to look up the index. Let me know if we can achieve this w/o making another look up